packages/muelu/src/Transfers/GeneralGeometric/MueLu_RegionRFactory_kokkos_def.hpp

```diff
@@ -0,0 +1,1199 @@
+    Set(coarseLevel, "Coordinates",   coarseCoordinates);
+    if(pL.get<bool>("keep coarse coords")) {
+      coarseLevel.Set<RCP<realvaluedmultivector_type> >("Coordinates2", coarseCoordinates, NoFactory::get());
+    }
+
+    R->SetFixedBlockSize(A->GetFixedBlockSize());
+    P->SetFixedBlockSize(A->GetFixedBlockSize());
+
+    Set(coarseLevel, "R", R);
+    Set(coarseLevel, "P", P);
+
+  } // Build()
+
+  template <class Scalar,class LocalOrdinal, class GlobalOrdinal, class Node>
+  void RegionRFactory_kokkos<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+  Build3D(const int numDimensions,
+          Teuchos::Array<LocalOrdinal>& lFineNodesPerDim,
+          const RCP<Matrix>& A,
+          const RCP<Xpetra::MultiVector<typename Teuchos::ScalarTraits<Scalar>::coordinateType, LocalOrdinal, GlobalOrdinal, Node> >& fineCoordinates,
+          RCP<Matrix>& R,
+          RCP<Xpetra::MultiVector<typename Teuchos::ScalarTraits<Scalar>::coordinateType, LocalOrdinal, GlobalOrdinal, Node> >& coarseCoordinates,
+          Teuchos::Array<LocalOrdinal>& lCoarseNodesPerDim) const {
+    using local_matrix_type = typename CrsMatrix::local_matrix_type;
+    using local_graph_type  = typename local_matrix_type::staticcrsgraph_type;
+    using row_map_type      = typename local_matrix_type::row_map_type::non_const_type;
+    using entries_type      = typename local_matrix_type::index_type::non_const_type;
+    using values_type       = typename local_matrix_type::values_type::non_const_type;
+
+    // Set debug outputs based on environment variable
+    RCP<Teuchos::FancyOStream> out;
+    if(const char* dbg = std::getenv("MUELU_REGIONRFACTORY_DEBUG")) {
+      out = Teuchos::fancyOStream(Teuchos::rcpFromRef(std::cout));
+      out->setShowAllFrontMatter(false).setShowProcRank(true);
+    } else {
+      out = Teuchos::getFancyOStream(rcp(new Teuchos::oblackholestream()));
+    }
+
+    // Now compute number of coarse grid points
+    for(int dim = 0; dim < numDimensions; ++dim) {
+      lCoarseNodesPerDim[dim] = lFineNodesPerDim[dim] / 3 + 1;
+    }
+    *out << "lCoarseNodesPerDim " << lCoarseNodesPerDim << std::endl;
+
+    // Grab the block size here and multiply all existing offsets by it
+    const LO blkSize = A->GetFixedBlockSize();
+    *out << "blkSize " << blkSize << std::endl;
+
+    // Based on lCoarseNodesPerDim and lFineNodesPerDim
+    // we can compute numRows, numCols and NNZ for R
+    const LO numRows = blkSize*lCoarseNodesPerDim[0]*lCoarseNodesPerDim[1]*lCoarseNodesPerDim[2];
+    const LO numCols = blkSize*lFineNodesPerDim[0]*lFineNodesPerDim[1]*lFineNodesPerDim[2];
+
+    // Create the coarse coordinates multivector
+    // so we can fill it on the fly while computing
+    // the restriction operator
+    RCP<Map> rowMap = MapFactory::Build(A->getRowMap()->lib(),
+                                        Teuchos::OrdinalTraits<GO>::invalid(),
+                                        numRows,
```

Comment:
It this correct? `numRows` is `numCoarseCoordinates*blkSize`, I think you might want `numCoarseCoordinates` here instead?

Reply:
That looks like something I introduced... :sweat_smile: 
This is used toward the bottom for the construction of R, however, and I think the use of `rowMap` is correct in that instance. I think `coarseCoordinates` is likely accidentally bigger than intended, but nothing else is impacted?